Add support for non blocking client hello callback (#2688)
aws/s2n-tls: docs/USAGE-GUIDE.md

typedef int s2n_client_hello_fn(struct s2n_connection *conn, void *ctx);
```

> The callback function takes as an input s2n-tls connection, which received

Author: done

ClientHello and context provided in **s2n_config_set_client_hello_cb**. The
callback can get any ClientHello information from the connection and use
**s2n_connection_set_config** call to change the config of the connection.

```c
int s2n_config_set_client_hello_cb_mode(struct s2n_config *config, s2n_client_hello_cb_mode cb_mode);
```
Sets the callback execution mode.

The callback can be be invoked in two modes
- **S2N_CLIENT_HELLO_CB_BLOCKING** (default):

>     In this mode s2n-tls expects the callback to complete it's work with

Author: fixed

    and return the appropriate response code after which handshake continues.
    If any of the properties of the connection were changed based on server_name
    extension the callback must return 1 or invoke **sn2_client_hello_server_name_used**,
    otherwise the callback can return 0 to continue handshake in s2n-tls.

- **S2N_CLIENT_HELLO_CB_NONBLOCKING**:

>     In non blocking mode for a response code of 0 from the callback **s2n_negotiate** will

Author: fixed

    return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status**
    will be set to **S2N_BLOCKED_ON_APPLICATION_INPUT**. The handshake is paused and further
    calls to **s2n_negotiate** will continue to return same error until **s2n_client_hello_cb_done**
    is invoked for the **s2n_connection**. This allows s2n-tls clients to process client_hello without
    blocking and resume the handshake at a later time.
    If any connection properties were changed on the basis of server_name extension then
    **s2n_client_hello_server_name_used** must be invoked before marking callback done.

The callback can return negative value to make s2n-tls terminate
handshake early with fatal handshake failure alert.

```c
int s2n_client_hello_cb_done(struct s2n_connection *conn)
```
Marks the non-blocking callback as complete

```c
int s2n_client_server_name_used(struct s2n_connection *conn)
```

> Indicates that connection properties were changed on the basis of server_name
> until the completion of the client_hello callback

Reviewer: ```suggestion
Indicates that connection properties were changed on the basis of server_name. Will trigger a
s2n-tls server to send the server_name extension. Must be called before s2n-tls finishes
processing the ClientHello.
```

typedef int s2n_client_hello_fn(struct s2n_connection *conn, void *ctx);
```

The callback function takes a s2n-tls connection as input, which receives the
ClientHello and the context previously provided in **s2n_config_set_client_hello_cb**.

> The callback can access any ClientHello information from the connection and use
> **s2n_connection_set_config** call to change the config of the connection.

Reviewer: ```suggestion
The callback can access any ClientHello information from the connection and use
the **s2n_connection_set_config** call to change the config of the connection.
```

Author: don't think `the` is needed here unless it was something like: `and use the **s2n_connection_set_config** method`

Reviewer: How does `method` vs `call` change things?

Author: my bad, didn't notice the `call` suffix :D updated

```c
int s2n_config_set_client_hello_cb_mode(struct s2n_config *config, s2n_client_hello_cb_mode cb_mode);
```
Sets the callback execution mode.

The callback can be be invoked in two modes
- **S2N_CLIENT_HELLO_CB_BLOCKING** (default):

    In this mode s2n-tls expects the callback to complete its work
    and return the appropriate response code before the handshake continues.

>     If any of the properties of the connection were changed based on server_name
>     extension the callback must either return 1 or invoke **sn2_client_hello_server_name_used**,

Reviewer: ```suggestion
    If any of the properties of the connection were changed based on the server_name
    extension the callback must either return 1 or invoke **s2n_client_hello_server_name_used**,
```

    otherwise the callback returns 0 to continue handshake.

- **S2N_CLIENT_HELLO_CB_NONBLOCKING**:

> 
>     In non-blocking mode, s2n-tls expects the callback to not complete it's work. If the callback

Reviewer: https://www.faa.gov/about/initiatives/plain_language/articles/its/
```suggestion

    In non-blocking mode, s2n-tls expects the callback to not complete its work. If the callback
```

    returns a response code of 0 s2n-tls will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED**
    error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**.

>     The handshake is paused and further calls to **s2n_negotiate** will continue to return
>     same error until **s2n_client_hello_cb_done** is invoked for the **s2n_connection** to resume

Reviewer: ```suggestion
    The handshake is paused and further calls to **s2n_negotiate** will continue to return the
    same error until **s2n_client_hello_cb_done** is invoked for the **s2n_connection** to resume
```

    the handshake. This allows s2n-tls clients to process client_hello without
    blocking and then resume the handshake at a later time.

>     If any connection properties were changed on the basis of server_name extension then
>     **s2n_client_hello_server_name_used** must be invoked before marking callback done.

Reviewer: ```suggestion
    If any connection properties were changed on the basis of the server_name extension then
    **s2n_client_hello_server_name_used** must be invoked before marking the callback done.
```

The callback can return a negative value to make s2n-tls terminate the
handshake early with a fatal handshake failure alert.

```c
int s2n_client_hello_cb_done(struct s2n_connection *conn)
```
Marks the non-blocking callback as complete

```c
int s2n_client_server_name_used(struct s2n_connection *conn)
```

> Indicates that connection properties were changed on the basis of server_name.
> Triggers a s2n-tls server to send the server_name extension. Must be called
> before s2n-tls finishes processing ClientHello.

Reviewer: ```suggestion
Indicates that connection properties were changed on the basis of server_name.
Triggers a s2n-tls server to send the server_name extension. Must be called
before s2n-tls finishes processing the ClientHello.
```

typedef int s2n_client_hello_fn(struct s2n_connection *conn, void *ctx);
```

The callback function takes a s2n-tls connection as input, which receives the
ClientHello and the context previously provided in **s2n_config_set_client_hello_cb**.
The callback can access any ClientHello information from the connection and use
the **s2n_connection_set_config** call to change the config of the connection.

```c
int s2n_config_set_client_hello_cb_mode(struct s2n_config *config, s2n_client_hello_cb_mode cb_mode);
```
Sets the callback execution mode.

The callback can be be invoked in two modes
- **S2N_CLIENT_HELLO_CB_BLOCKING** (default):

    In this mode s2n-tls expects the callback to complete its work
    and return the appropriate response code before the handshake continues.
    If any of the connection properties were changed based on the server_name

>     extension the callback must either return 1 or invoke **s2n_client_hello_server_name_used**,

Author: yes, will update
